docs/Technical-Documentation/tech-memos/feature-toggle-system.md

```diff
@@ -0,0 +1,620 @@
+- `SET_FEATURE_FLAGS` - Store fetched flags
+- `SET_FEATURE_FLAGS_ERROR` - Store error
+- `CLEAR_FEATURE_FLAGS_ERROR` - Clear error
+- `RESET_FEATURE_FLAGS` - Clear all flags (on logout)
+
+**State shape**:
+```javascript
+{
+  featureFlags: {
+    flags: {
+      pia_submission: { enabled: true, config: {...} }
+    },
+    loading: false,
+    error: null,
+    lastFetched: 1700251234567,
+  }
+}
+```
+
+**Example Thunk**:
+```javascript
+export const fetchFeatureFlags = () => async (dispatch) => {
+  dispatch({ type: FETCH_FEATURE_FLAGS })
+  try {
+    const { data } = await axiosInstance.get(`${BACKEND_URL}/v1/features/`)
+    dispatch({ type: SET_FEATURE_FLAGS, payload: { flags: data } })
+  } catch (error) {
+    dispatch({ type: SET_FEATURE_FLAGS_ERROR, payload: { error } })
+  }
+}
+```
+
+### Auth Check Integration
+
+Because the feature flags are dependent on the auth state of the user, it makes sense to couple the execution of the `fetchAuth`
+with the handling of the feature flags. That is, when a user authenticates successfully, we should immediately fetch the feature
+flags and populate the redux store. Conversely, if the user does not successfully authenticate we should ensure the feature state
+is cleared. Because the auth check also runs via the `IdleTimer` we can guarantee a consistent and timely update to the feature
+flag state.
+
+**Location**: `tdrs-frontend/src/actions/auth.js`
+
+```javascript
+export const fetchAuth = () => async (dispatch) => {
+  dispatch({ type: FETCH_AUTH })
+  // ...Existing code
+  if (data?.user) {
+    // ...Existing code
+    // Update feature flags
+    dispatch(fetchFeatureFlags())
+  } else {
+    // ...Existing code
+    // Clear feature flags
+    dispatch(resetFeatureFlags())
+  }
+}
+```
+
+### Periodic Refresh Strategy
+
+In the event that the this pattern becomes more frequent, or the implementer decides not to couple the auth check with the
+feature flag update. This technical memorandum suggest a generic data component that will handle the interval based redux
+updates.
+
+**Example Component**:
+```javascript
+const ReduxRefresh = ({ thunk, intervalMs = 300000 }) => {
+  const dispatch = useDispatch()
+  const isAuthenticated = useSelector((state) => state.auth.authenticated)
+
+  useEffect(() => {
+    if (!isAuthenticated) return // Should probably clear redux state here
+
+    const interval = setInterval(() => {
+      dispatch(thunk)
+    }, intervalMs)
+
+    return () => clearInterval(interval)
+  }, [isAuthenticated, intervalMs, dispatch])
+
+  return null
+}
+```
+
+### Frontend React Hooks
+
+Continuing the frontend's effort to move re-useable state and logic to hooks, a `useFeatureFlag` hook is suggested to allow
+components easy abstracted access to the feature flags redux state.
+
+**Location**: `tdrs-frontend/src/hooks/useFeatureFlag.js`
+
+```javascript
+import { useSelector } from 'react-redux';
+
+export const useFeatureFlag = (featureName) => {
```

Comment:
I think calling a useSelector inside a nested function that is being conditionally called is not right. here: https://legacy.reactjs.org/docs/hooks-rules.html